File: week-4/assignment-4/python/src/scheduler.py
import random
from models import Employee
from config import config

DAYS = config["days"]
SHIFTS = config["shifts"]


class Scheduler:
    def __init__(self):
        """
        Initialize scheduler with empty employee and schedule structures.
        Also validates the configuration for required keys.
        """
        self.employees = {}
        self.schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}

    def add_employee(self, name, preferences):
        """
        Add an employee with shift preferences.

        Args:
            name (str): Name of the employee.
            preferences (list): List of preferred shifts for the employee.
        """
        self.employees[name] = Employee(name, preferences)

    def get_schedule(self):
        """
        Retrieve the current shift schedule.

        Returns:
            dict: Nested dict of days and their corresponding shifts and assigned employee names.
        """
        return self.schedule

    def get_employees(self):
        """
        Retrieve the list of employees.

        Returns:
            dict: Dictionary of employee names to Employee objects.
        """
        return self.employees

    def reset_schedule_and_employee_counters(self):
        """Reset the shift schedule and all employee assignment counters."""
        self.schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}
        for employee in self.employees.values():
            employee.assigned_days = set()

    def get_eligible_employees_for_day(self, day):
        """
        Get a list of employees eligible to work on a given day.

        Args:
            day (str): The day to check availability for.

        Returns:
            list: List of eligible Employee objects.
        """
        return [
            employee
            for employee in self.employees.values()
            if len(employee.assigned_days) < 5 and day not in employee.assigned_days
        ]

    def sorted_employees_by_preference_and_days_assigned(self, day, shift):
        """
        Get eligible employees for a shift on a day, sorted by fewest days assigned and preference rank.

        Args:
            day (str): The day to assign.
            shift (str): The shift name.

        Returns:
            list: Sorted list of tuples (days_assigned, preference_rank, Employee)
        """
        available_employees = [
            (
                len(employee.assigned_days),
                employee.preference_rank.get(shift, 99),
                employee,
            )
            for employee in self.get_eligible_employees_for_day(day)
        ]
        # Sort by days_assigned ascending, then preference_rank ascending
        return sorted(available_employees, key=lambda x: (x[0], x[1]))

    def assign_employee(self, employee, day, shift):
        """
        Assign a specific employee to a shift on a given day.

        Args:
            employee (Employee): The Employee object.
            day (str): The day of assignment.
            shift (str): The shift name.
        """
        self.schedule[day][shift].append(employee.name)
        employee.assigned_days.add(day)

    def assign_shifts(self):
        """Assign employees to shifts while respecting preferences and constraints."""
        self.reset_schedule_and_employee_counters()

        for day_index, day in enumerate(DAYS):
            for shift in SHIFTS:
                available_employees = (
                    self.sorted_employees_by_preference_and_days_assigned(day, shift)
                )

                # Assign top preferred employees
                while len(self.schedule[day][shift]) < 2 and available_employees:
                    chosen_employee = available_employees.pop(0)[2]
                    self.assign_employee(chosen_employee, day, shift)

                # If still unfilled, assign random eligible employees
                if len(self.schedule[day][shift]) < 2:
                    eligible_employees = self.get_eligible_employees_for_day(day)
                    random.shuffle(eligible_employees)
                    for employee in eligible_employees:
                        if len(self.schedule[day][shift]) < 2:
                            self.assign_employee(employee, day, shift)

            # Handle any unassigned employees for this day
            unassigned_employees = self.get_eligible_employees_for_day(day)

            for employee in unassigned_employees:
                shift_assigned = False
                for shift in employee.preferences:
                    if len(self.schedule[day][shift]) < 2:
                        self.assign_employee(employee, day, shift)
                        shift_assigned = True
                        break

                if not shift_assigned:
                    # Try next day's preferred shifts if current day is full
                    for offset in range(1, len(DAYS)):
                        next_day = DAYS[(day_index + offset) % len(DAYS)]
                        for shift in employee.preferences:
                            if (
                                len(self.schedule[next_day][shift]) < 2
                                and next_day not in employee.assigned_days
                            ):
                                self.assign_employee(employee, next_day, shift)
                                shift_assigned = True
                                break
                        if shift_assigned:
                            break

File: week-4/assignment-4/go/scheduler/scheduler.go
package scheduler

import (
	"math/rand"
	"sort"
	"time"

	"github.com/aashishshrestha09/employee-shift-scheduler-go/config"
	"github.com/aashishshrestha09/employee-shift-scheduler-go/models"
)

// Scheduler contains employee data and the generated schedule
type Scheduler struct {
	Employees map[string]*models.Employee
	Schedule  map[string]map[string][]string
}

var (
	Days   = config.Days
	Shifts = config.Shifts
)

// NewScheduler initializes a Scheduler with empty employee list and schedule map
func NewScheduler() *Scheduler {
	scheduler := &Scheduler{
		Employees: make(map[string]*models.Employee),
		Schedule:  make(map[string]map[string][]string),
	}
	// Initialize schedule map
	for _, day := range Days {
		scheduler.Schedule[day] = make(map[string][]string)
		for _, shift := range Shifts {
			scheduler.Schedule[day][shift] = []string{}
		}
	}
	return scheduler
}

// GetSchedule returns the full current schedule mapping days and shifts to assigned employees
func (s *Scheduler) GetSchedule() map[string]map[string][]string {
	return s.Schedule
}

// GetEmployees returns the map of all employees managed by the scheduler
func (s *Scheduler) GetEmployees() map[string]*models.Employee {
	return s.Employees
}

// GetEmployeePreferences returns the shift preferences of an employee by name
func (s *Scheduler) GetEmployeePreferences(name string) []string {
	if employee, ok := s.Employees[name]; ok {
		return employee.Preferences
	}
	return nil
}

// AddEmployee adds a new employee with their ordered shift preferences
func (s *Scheduler) AddEmployee(name string, preferences []string) {
	rank := make(map[string]int)
	for i, shift := range preferences {
		rank[shift] = i
	}
	employee := &models.Employee{
		Name:           name,
		Preferences:    preferences,
		PreferenceRank: rank,
		AssignedDays:   make(map[string]bool),
	}
	s.Employees[name] = employee
}

// ResetScheduleAndCounters clears the current schedule and resets employee assignment counters
func (s *Scheduler) ResetScheduleAndCounters() {
	for _, day := range Days {
		for _, shift := range Shifts {
			s.Schedule[day][shift] = []string{}
		}
	}
	for _, employee := range s.Employees {
		employee.AssignedDays = make(map[string]bool)
	}
}

// GetEligibleEmployeesForDay returns employees who have less than 5 assigned days and are free on the given day
func (s *Scheduler) GetEligibleEmployeesForDay(day string) []*models.Employee {
	eligible := []*models.Employee{}
	for _, employee := range s.Employees {
		if len(employee.AssignedDays) < 5 && !employee.AssignedDays[day] {
			eligible = append(eligible, employee)
		}
	}
	return eligible
}

// employeeSortHelper is used for sorting employees by days assigned and preference rank
type employeeSortHelper struct {
	DaysAssigned   int
	PreferenceRank int
	Employee       *models.Employee
}

// SortedEmployeesByPreferenceAndDaysAssigned returns employees eligible for a day sorted by fewest days assigned and best preference rank for the shift
func (s *Scheduler) SortedEmployeesByPreferenceAndDaysAssigned(day string, shift string) []employeeSortHelper {
	availableEmployees := []employeeSortHelper{}
	for _, employee := range s.GetEligibleEmployeesForDay(day) {
		rank, ok := employee.PreferenceRank[shift]
		if !ok {
			rank = 99 // Low preference if not in preferences
		}
		availableEmployees = append(availableEmployees, employeeSortHelper{
			DaysAssigned:   len(employee.AssignedDays),
			PreferenceRank: rank,
			Employee:       employee,
		})
	}

	sort.Slice(availableEmployees, func(i, j int) bool {
		if availableEmployees[i].DaysAssigned == availableEmployees[j].DaysAssigned {
			return availableEmployees[i].PreferenceRank < availableEmployees[j].PreferenceRank
		}
		return availableEmployees[i].DaysAssigned < availableEmployees[j].DaysAssigned
	})

	return availableEmployees
}

// AssignEmployee assigns an employee to a specific day and shift and marks the day as assigned for that employee
func (s *Scheduler) AssignEmployee(employee *models.Employee, day string, shift string) {
	s.Schedule[day][shift] = append(s.Schedule[day][shift], employee.Name)
	employee.AssignedDays[day] = true
}

// AssignShifts runs the scheduling algorithm to assign employees to shifts across the week
func (s *Scheduler) AssignShifts() {
	s.ResetScheduleAndCounters()

	rand.Seed(time.Now().UnixNano())

	for dayIndex, day := range Days {
		for _, shift := range Shifts {

			// Get employees sorted by fewest assigned days and preference for this shift
			availableEmployees := s.SortedEmployeesByPreferenceAndDaysAssigned(day, shift)

			// Assign top preferred employees until 2 per shift
			for len(s.Schedule[day][shift]) < 2 && len(availableEmployees) > 0 {
				employee := availableEmployees[0].Employee
				availableEmployees = availableEmployees[1:]
				s.AssignEmployee(employee, day, shift)
			}

			// If still unfilled, assign random eligible employees
			if len(s.Schedule[day][shift]) < 2 {
				eligibleEmployees := s.GetEligibleEmployeesForDay(day)
				rand.Shuffle(len(eligibleEmployees), func(i, j int) {
					eligibleEmployees[i], eligibleEmployees[j] = eligibleEmployees[j], eligibleEmployees[i]
				})
				for _, employee := range eligibleEmployees {
					if len(s.Schedule[day][shift]) < 2 {
						s.AssignEmployee(employee, day, shift)
					}
				}
			}
		}

		// Handle unassigned employees for the day
		unassignedEmployees := s.GetEligibleEmployeesForDay(day)

		for _, employee := range unassignedEmployees {
			shiftAssigned := false
			for _, preferenceShift := range employee.Preferences {
				if len(s.Schedule[day][preferenceShift]) < 2 {
					s.AssignEmployee(employee, day, preferenceShift)
					shiftAssigned = true
					break
				}
			}

			if !shiftAssigned {
				// Try next days shifts if current day is full
				for offset := 1; offset < len(Days); offset++ {
					nextDay := Days[(dayIndex+offset)%len(Days)]
					for _, preferenceShift := range employee.Preferences {
						if len(s.Schedule[nextDay][preferenceShift]) < 2 && !employee.AssignedDays[nextDay] {
							s.AssignEmployee(employee, nextDay, preferenceShift)
							shiftAssigned = true
							break
						}
					}
					if shiftAssigned {
						break
					}
				}
			}
		}
	}
}
